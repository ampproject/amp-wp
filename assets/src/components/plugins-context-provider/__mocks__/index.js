/**
 * External dependencies
 */
import PropTypes from 'prop-types';

/**
 * WordPress dependencies
 */
import { createContext } from '@wordpress/element';

export const Plugins = createContext();

/**
 * MOCK.
 *
 * @param {Object}  props
 * @param {any}     props.children
 */
export function PluginsContextProvider( { children } ) {
	return (
		<Plugins.Provider value={
			{
				fetchingPlugins: false,
				plugins: [],
			}
		}>
			{ children }
		</Plugins.Provider>
	);
}
PluginsContextProvider.propTypes = {
	children: PropTypes.any,
};
