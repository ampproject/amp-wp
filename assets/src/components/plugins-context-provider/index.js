/**
 * External dependencies
 */
import PropTypes from 'prop-types';

/**
 * WordPress dependencies
 */
import {
	createContext,
	useContext,
	useEffect,
	useRef,
	useState,
} from '@wordpress/element';
import apiFetch from '@wordpress/api-fetch';

/**
 * Internal dependencies
 */
import { ErrorContext } from '../error-context-provider';
import { useAsyncError } from '../../utils/use-async-error';

export const Plugins = createContext();

/**
 * Plugins context provider.
 *
 * @param {Object} props Component props.
 * @param {any} props.children Component children.
 * @param {boolean} props.hasErrorBoundary Whether the component is wrapped in an error boundary.
 */
export function PluginsContextProvider( {
	children,
	hasErrorBoundary = false,
} ) {
	const [ plugins, setPlugins ] = useState();
	const [ fetchingPlugins, setFetchingPlugins ] = useState( null );

	const { error, setError } = useContext( ErrorContext );
	const { setAsyncError } = useAsyncError();

	/**
	 * This component sets state inside async functions.
	 * Use this ref to prevent state updates after unmount.
	 */
	const hasUnmounted = useRef( false );
	useEffect( () => () => {
		hasUnmounted.current = true;
	}, [] );

	/**
	 * Fetches validated URL data.
	 */
	useEffect( () => {
		if ( error || plugins || fetchingPlugins ) {
			return;
		}

		( async () => {
			setFetchingPlugins( true );

			try {
				const fetchedPlugins = await apiFetch( {
					path: '/wp/v2/plugins',
				} );

				if ( hasUnmounted.current === true ) {
					return;
				}

				setPlugins( fetchedPlugins );
			} catch ( e ) {
				if ( hasUnmounted.current === true ) {
					return;
				}

				setError( e );

				if ( hasErrorBoundary ) {
					setAsyncError( e );
				}

				return;
			}

			setFetchingPlugins( false );
		} )();
	}, [ error, fetchingPlugins, hasErrorBoundary, plugins, setAsyncError, setError ] );

	return (
		<Plugins.Provider
			value={ {
				fetchingPlugins,
				plugins,
			} }
		>
			{ children }
		</Plugins.Provider>
	);
}
PluginsContextProvider.propTypes = {
	children: PropTypes.any,
	hasErrorBoundary: PropTypes.bool,
};
