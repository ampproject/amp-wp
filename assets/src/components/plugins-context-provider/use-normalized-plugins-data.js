/**
 * WordPress dependencies
 */
import { useContext, useEffect, useState } from '@wordpress/element';

/**
 * Internal dependencies
 */
import { getPluginSlugFromFile } from '../../common/helpers/get-plugin-slug-from-file';
import { Plugins } from './index';

export function useNormalizedPluginsData() {
	const { fetchingPlugins, plugins } = useContext(Plugins);
	const [normalizedPluginsData, setNormalizedPluginsData] = useState({});

	useEffect(() => {
		if (fetchingPlugins || plugins.length === 0) {
			return;
		}

		setNormalizedPluginsData(
			plugins.reduce((accumulatedPluginsData, source) => {
				const slug = getPluginSlugFromFile(source.plugin);

				return {
					...accumulatedPluginsData,
					[slug]: Object.keys(source).reduce(
						(props, key) => ({
							...props,
							slug,
							// Flatten every prop that contains a `raw` member.
							[key]: source[key]?.raw ?? source[key],
						}),
						{}
					),
				};
			}, {})
		);
	}, [fetchingPlugins, plugins]);

	return normalizedPluginsData;
}
