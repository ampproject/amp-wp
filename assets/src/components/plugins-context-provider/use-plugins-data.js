/**
 * WordPress dependencies
 */
import { useContext, useEffect, useState } from '@wordpress/element';
import { __, sprintf } from '@wordpress/i18n';

/**
 * Internal dependencies
 */
import { Plugins } from './index';

export default function usePluginsData() {
	const { plugins, fetchingPlugins } = useContext( Plugins );
	const [ nameBySlug, setNameBySlug ] = useState( [] );

	useEffect( () => {
		if ( fetchingPlugins || ! plugins ) {
			return;
		}

		setNameBySlug( () => plugins.reduce( ( acc, plugin ) => {
			const slug = plugin.plugin.split( '/' )[ 0 ];
			return {
				...acc,
				[ slug ]: plugin.name,
			};
		}, {} ) );
	}, [ fetchingPlugins, plugins ] );

	const getPluginNameBySlug = ( slug ) => nameBySlug[ slug ] ??
		sprintf(
			/* translators: Plugin slug. */
			__( 'Plugin: %s', 'amp' ),
			slug,
		);

	return {
		getPluginNameBySlug,
		plugins,
	};
}
