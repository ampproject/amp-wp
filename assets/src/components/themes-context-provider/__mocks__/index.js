/**
 * External dependencies
 */
import PropTypes from 'prop-types';

/**
 * WordPress dependencies
 */
import { createContext } from '@wordpress/element';

export const Themes = createContext();

/**
 * MOCK.
 *
 * @param {Object}  props
 * @param {any}     props.children
 */
export function ThemesContextProvider( { children } ) {
	return (
		<Themes.Provider value={
			{
				fetchingThemes: false,
				themes: [],
			}
		}>
			{ children }
		</Themes.Provider>
	);
}
ThemesContextProvider.propTypes = {
	children: PropTypes.any,
};
