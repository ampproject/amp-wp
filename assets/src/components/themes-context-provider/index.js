/**
 * External dependencies
 */
import PropTypes from 'prop-types';

/**
 * WordPress dependencies
 */
import {
	createContext,
	useContext,
	useEffect,
	useRef,
	useState,
} from '@wordpress/element';
import apiFetch from '@wordpress/api-fetch';

/**
 * Internal dependencies
 */
import { ErrorContext } from '../error-context-provider';
import { useAsyncError } from '../../utils/use-async-error';

export const Themes = createContext();

/**
 * Themes context provider.
 *
 * @param {Object} props Component props.
 * @param {any} props.children Component children.
 * @param {boolean} props.hasErrorBoundary Whether the component is wrapped in an error boundary.
 */
export function ThemesContextProvider( {
	children,
	hasErrorBoundary = false,
} ) {
	const [ themes, setThemes ] = useState();
	const [ fetchingThemes, setFetchingThemes ] = useState( null );

	const { error, setError } = useContext( ErrorContext );
	const { setAsyncError } = useAsyncError();

	/**
	 * This component sets state inside async functions.
	 * Use this ref to prevent state updates after unmount.
	 */
	const hasUnmounted = useRef( false );
	useEffect( () => () => {
		hasUnmounted.current = true;
	}, [] );

	/**
	 * Fetches validated URL data.
	 */
	useEffect( () => {
		if ( error || themes || fetchingThemes ) {
			return;
		}

		( async () => {
			setFetchingThemes( true );

			try {
				const fetchedThemes = await apiFetch( {
					path: '/wp/v2/themes',
				} );

				if ( hasUnmounted.current === true ) {
					return;
				}

				setThemes( fetchedThemes );
			} catch ( e ) {
				if ( hasUnmounted.current === true ) {
					return;
				}

				setError( e );

				if ( hasErrorBoundary ) {
					setAsyncError( e );
				}

				return;
			}

			setFetchingThemes( false );
		} )();
	}, [ error, fetchingThemes, hasErrorBoundary, themes, setAsyncError, setError ] );

	return (
		<Themes.Provider
			value={ {
				fetchingThemes,
				themes,
			} }
		>
			{ children }
		</Themes.Provider>
	);
}
ThemesContextProvider.propTypes = {
	children: PropTypes.any,
	hasErrorBoundary: PropTypes.bool,
};
