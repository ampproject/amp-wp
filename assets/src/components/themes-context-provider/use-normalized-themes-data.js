/**
 * WordPress dependencies
 */
import { useContext, useEffect, useState } from '@wordpress/element';

/**
 * Internal dependencies
 */
import { Themes } from './index';

export function useNormalizedThemesData() {
	const { fetchingThemes, themes } = useContext(Themes);
	const [normalizedThemesData, setNormalizedThemesData] = useState({});

	useEffect(() => {
		if (fetchingThemes || themes.length === 0) {
			return;
		}

		setNormalizedThemesData(() =>
			themes.reduce(
				(accumulatedThemesData, source) => ({
					...accumulatedThemesData,
					[source.stylesheet]: Object.keys(source).reduce(
						(props, key) => {
							const normalizedData = {
								...props,
								slug: source.stylesheet,
								// Flatten every prop that contains a `raw` member.
								[key]: source[key]?.raw ?? source[key],
							};

							// Look for a child theme.
							const childTheme = themes.find(
								(theme) =>
									theme.template === source.stylesheet &&
									theme.template !== theme.stylesheet
							)?.stylesheet;
							if (childTheme) {
								normalizedData.child = childTheme;
							}

							// Set parent theme if we're dealing with a child theme.
							if (
								source.template &&
								source.template !== source.stylesheet
							) {
								normalizedData.parent = source.template;
							}

							return normalizedData;
						},
						{}
					),
				}),
				{}
			)
		);
	}, [fetchingThemes, themes]);

	return normalizedThemesData;
}
