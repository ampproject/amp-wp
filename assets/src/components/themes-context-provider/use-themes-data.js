/**
 * WordPress dependencies
 */
import { useContext, useEffect, useState } from '@wordpress/element';
import { __, sprintf } from '@wordpress/i18n';

/**
 * Internal dependencies
 */
import { Themes } from './index';

export default function useThemesData() {
	const { themes, fetchingThemes } = useContext( Themes );
	const [ nameBySlug, setNameBySlug ] = useState( [] );

	useEffect( () => {
		if ( fetchingThemes || ! themes ) {
			return;
		}

		setNameBySlug( () => themes.reduce( ( acc, theme ) => ( {
			...acc,
			[ theme.stylesheet ]: theme.name.raw,
		} ), {} ) );
	}, [ fetchingThemes, themes ] );

	const getThemeNameBySlug = ( slug ) => nameBySlug[ slug ] ??
		sprintf(
			/* translators: Theme slug. */
			__( 'Theme: %s', 'amp' ),
			slug,
		);

	return {
		getThemeNameBySlug,
		themes,
	};
}
