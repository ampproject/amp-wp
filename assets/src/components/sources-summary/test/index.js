/**
 * WordPress dependencies
 */
import { render } from '@wordpress/element';

/**
 * External dependencies
 */
import { act } from 'react-dom/test-utils';

/**
 * Internal dependencies
 */
import SourcesSummary from '..';
import { PluginsContextProvider } from '../../plugins-context-provider';
import { ThemesContextProvider } from '../../themes-context-provider';

jest.mock( '../../plugins-context-provider' );
jest.mock( '../../themes-context-provider' );

let container;

describe( 'SourcesSummary', () => {
	beforeEach( () => {
		container = document.createElement( 'div' );
		document.body.appendChild( container );
	} );

	afterEach( () => {
		document.body.removeChild( container );
		container = null;
	} );

	it( 'renders nothing if no sources array is provided', () => {
		act( () => {
			render(
				<SourcesSummary sources={ [] } />,
				container,
			);
		} );

		expect( container.textContent ).toBe( '' );
	} );

	it( 'renders validated theme name if no sources are provided', () => {
		act( () => {
			render(
				<PluginsContextProvider>
					<ThemesContextProvider>
						<SourcesSummary validatedTheme="foo" />
					</ThemesContextProvider>
				</PluginsContextProvider>,
				container,
			);
		} );

		expect( container.innerHTML ).toMatchSnapshot();
		expect( container.textContent ).toContain( 'foo' );
	} );

	it( 'renders a list of sources', () => {
		act( () => {
			render(
				<PluginsContextProvider>
					<ThemesContextProvider>
						<SourcesSummary sources={ [
							{ type: 'plugin', name: 'foo' },
							{ type: 'theme', name: 'bar' },
						] } />
					</ThemesContextProvider>
				</PluginsContextProvider>,
				container,
			);
		} );

		expect( container.innerHTML ).toMatchSnapshot();
		expect( container.textContent ).toContain( 'foo' );
		expect( container.textContent ).toContain( 'bar' );
	} );
} );
