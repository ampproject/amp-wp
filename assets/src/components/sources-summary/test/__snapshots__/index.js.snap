// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SourcesSummary renders a list of sources 1`] = `"<strong class=\\"source\\"><span class=\\"dashicon dashicons dashicons-admin-plugins\\"></span>Plugin: foo</strong><strong class=\\"source\\"><span class=\\"dashicon dashicons dashicons-admin-appearance\\"></span>Theme: bar</strong>"`;

exports[`SourcesSummary renders validated theme name if no sources are provided 1`] = `"<strong class=\\"source\\"><span class=\\"dashicon dashicons dashicons-admin-appearance\\"></span>Theme: foo</strong>"`;
